Added the Tiberian Sun Nod Obelisk charge animation

Mailaender opened the pull request:
I was forced to do a lot of cleanup, but that isn't necessarily a bad thing.

Mailaender committed:
separate with newlines for readability

Mailaender committed:
readonly

pchote commented:
Conflicts with #7980.

Mailaender commented:
You mean the other way round I guess.

pchote commented:
No. #7980 was written and filed before this.

pchote commented on OpenRA.Mods.Common/Traits/Air/AttackPlane.cs:
```diff
@@ -37,7 +37,7 @@ public override Activity GetAttackActivity(Actor self, Target newTarget, bool al
 		}
 
-		protected override bool CanAttack(Actor self, Target target)
+		public override bool CanAttack(Actor self, Target target)
```
This only seems to be necessary because you split out the internal activities. Why?

Mailaender replied:
Because they were activities inside the traits namespace.

pchote replied:
I'm not really happy with the idea of mixing an architectural rework (which I personally don't think should even be changed) like this in with new features.

Mailaender replied:
Removed it again. I was foreseeing someone will make me as the reviewers have a tendency to "oh now you touched that legacy code slightly how about redoing it completely" and that is more painful to do afterwards.

pchote commented on OpenRA.Mods.Common/Traits/Sound/ChargeSound.cs:
```diff
@@ -0,0 +1,36 @@
+	public class ChargeSoundInfo : ITraitInfo
+	{
+		[Desc("Sound to play when actor charges.")]
+		public readonly string ChargeAudio = null;
```
I think this would make more sense as part of `AttackCharge`.

Mailaender replied:
Moved.

Mailaender commented:
Rebased.

Mailaender commented:
Rebased again as https://github.com/OpenRA/OpenRA/pull/7980 has been merged.

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackCharge.cs:
```diff
@@ -29,30 +29,33 @@ class AttackChargeInfo : AttackOmniInfo
 	class AttackCharge : AttackOmni, ITick, INotifyAttack, ISync
 	{
-		readonly AttackChargeInfo info;
+		public readonly new AttackChargeInfo Info;
```
Please see the earlier discussions about not using new` in this situation.

Mailaender replied:
What earlier discussion? What to use instead? I just did what the compiler told me.

Mailaender committed:
split render building/overlay and play sound for charge

to add the Tiberian Sun Nod Obelisk

tick slightly faster

tweak the local offsets

Mailaender commented:
Updated.

abcdefg30 commented on mods/ts/sequences/structures.yaml:
```diff
@@ -582,11 +582,9 @@ naobel:
-#		Tick: 240
+	active: ntobel_b
+		Length: 12
+		Tick: 240
```
I think `200` would suit better.

abcdefg30 commented:
Otherwise :+1: / :white_check_mark:

pchote commented on OpenRA.Mods.TS/Traits/Render/WithChargeOverlay.cs:
```diff
@@ -0,0 +1,58 @@
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.TS.Traits
```
This probably wants to live in Mods.Common.

Mailaender replied:
I don't think so.

pchote replied:
There was a discussion in IRC about this between myself, @reaperrr, and @Phrohdoh. That has probably been permanently lost now, but they may want to weigh in with their own thoughts.

Basically: our criteria for including things in the common dll is "this is a general purpose trait" and not just "this is used in more than one of our shipping mods".  This trait is the overlay counterpart to `RenderBuildingCharge`, and so it belongs in the same namespace as that trait (which is Mods.Common).

Mailaender replied:
That makes sense. Moved.

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -28,6 +28,7 @@ public interface IPlaceBuildingDecoration
 	}
 
 	public interface INotifyAttack { void Attacking(Actor self, Target target, Armament a, Barrel barrel); }
+	public interface INotifyCharging { void Charging(Actor self); }
```
This interface would be more flexible if it also provided the target that the actor is charging to attack.

pchote commented on OpenRA.Mods.TS/Traits/Render/WithChargeOverlay.cs:
```diff
@@ -0,0 +1,58 @@
+		public object Create(ActorInitializer init) { return new WithChargeOverlay(init, this); }
+	}
+
+	public class WithChargeOverlay : INotifyCharging
```
Modders (and maybe even the RA2 prism towers?) will want support for damaged charging overlays.  I suggest implementing `INotifyDamageStateChanged` now (see `WithIdleOverlay` for more details).

pchote commented on mods/ts/rules/nod-support.yaml:
```diff
@@ -88,16 +88,18 @@ NAOBEL:
 	Armament:
 		Weapon: LaserFire
-		LocalOffset: 0,0,725
+		LocalOffset: 1500,250,1250
```
This is a bit too high and too the right:

![screen shot 2015-05-02 at 12 03 21 2](https://cloud.githubusercontent.com/assets/167819/7439047/5d814c26-f0c3-11e4-94ba-cb93ed0997fc.png)

pchote commented:
:+1: after fixing those simple issues.

Mailaender commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithChargeOverlay.cs:
```diff
@@ -0,0 +1,62 @@
+		public void Charging(Actor self, Target target)
+		{
+			charging = true;
+			overlay.PlayThen(info.Sequence, () => charging = false);
```
This also needs to use NormalizeSequence.

Mailaender commented:
Updated. Also fixed an NRE caused by these changes.

penev92 commented:
- If you sell the TS Obelisk while it is charging, the crystal on the top will remain in its place and play the charge animation while the sell animation is playing.
 - If you sell or destroy the TS Obelisk while it is charging, the charge sound will still be played (I guess there is no way to cancel it though)

I couldn't spot any of those in RA or TD. My guess is the charge sound in RA is too short, so the issue isn't noticeable, and the `RenderBuildingCharge` changes the charge animation with the sell animation when needed. It would be nice to have `WithChargeOverlay` disable the charge animation in that case as well.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithChargeOverlay.cs:
```diff
@@ -0,0 +1,62 @@
+
+			renderSprites = init.Self.Trait<RenderSprites>();
+
+			overlay = new Animation(init.World, renderSprites.GetImage(init.Self));
```
You need a NormalizeSequence here too, otherwise it will display the wrong sequence if it is spawned damaged.

Mailaender commented:
Fixed the selling glitch.

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/0652f76a7f36ce8050d1f54636e173095ab886d0...279ac70e138dbdde5613d1c5ea2e6b7609b86287)
